The Application Gateway should remove the x-forwarded-proto header (#3391)
kyma-project/kyma: docs/application-connector/docs/02-03-proxy-service.md

>**NOTE:** The system creates an Access Service for every external API registered by the Application Registry.

The following diagram illustrates how the Application Gateway interacts with other components and external APIs

> which might be unsecured or secured with various security mechanism and also protected against cross-site request forgery (CSRF) attacks.

Reviewer: ```suggestion
which are either unsecured or secured with various security mechanisms and protected against cross-site request forgery (CSRF) attacks.
```

![Application Gateway Diagram](./assets/003-architecture-proxy-service.svg)

1. A lambda function calls the Access Service. The name of every Access Service follows this format: `app-{application-name}-{service-id}`
2. The Access Service exposes the Application Gateway.
3. The Application Gateway extracts the Application name and the service ID from the name of the Access Service name. Using the extracted Application name, the Application Gateway finds the respective Application custom resource and obtains the information about the registered external API, such as the API URL and security credentials.
4. The Application Gateway gets a token from the OAuth server.

> 5. The Application Gateway gets a CSRF token from the endpoint exposed by the upstream service. (optional step based on API registration in App Registry)

Reviewer: ```suggestion
5. The Application Gateway gets a CSRF token from the endpoint exposed by the upstream service. This step is optional and is valid only for the API registration in the Application Registry.
```

Author: It is not exactly what I was trying to express:
>This step is optional and is valid only for the API which was registered with CSRF token turned on.
Something like that is more accurate IMO.

6. The Application Gateway calls the target API.

## Caching

To ensure optimal performance, the Application Gateway caches the OAuth tokens and CSRF tokens it obtains. If the service doesn't find valid tokens for the call it makes, it gets new tokens from the OAuth server and the CSRF token endpoint.
Additionally, the service caches ReverseProxy objects used to proxy the requests to the underlying URL.

## Handling of headers

> Please note that Application Gateway is removing the following headers while making the calls to registered applications:

Reviewer: ```suggestion
The Application Gateway removes the following headers while making calls to the registered applications:
```

> - X-Forwarded-Proto

Reviewer: ```suggestion
- `X-Forwarded-Proto`
```

> - X-Forwarded-For

Reviewer: ```suggestion
- `X-Forwarded-For`
```

> - X-Forwarded-Host

Reviewer: ```suggestion
- `X-Forwarded-Host`
```

> - X-Forwarded-Client-Cert

Reviewer: ```suggestion
- `X-Forwarded-Client-Cert`
```

> In addition, the header `User-Agent` is set to empty value if present.

Reviewer: ```suggestion
In addition, the `User-Agent` header is used for empty values.
```

> In addition, the header `User-Agent` is set to empty value if present.

Reviewer 2: Not really, if User-Agent header is set then we do not touch it. If it is not set then we set it to an empty value to stop gateway setting it to the default value.

Author: Ok, good catch.

Author: >In addition, the header `User-Agent` is set to empty value if present. In addition, the header `User-Agent` is set to empty value if it is not given in the call. It will prevent setting it to the default value.

> In addition, the header `User-Agent` is set to empty value if it is not given in the call. It will prevent setting it to the default value.

Reviewer: ```suggestion
In addition, the `User-Agent` header is set to an empty value not specified in the call, which prevents from setting a default value.
```
